lxd/device/disk.go

```diff
@@ -138,8 +138,8 @@ func (d *disk) validateConfig(instConf instance.ConfigReader) error {
 	}
 
 	// QMP allows node names to be 31 characters. LXD prefixes any block device with `lxd_` which leaves the user with 27 characters.
-	if len(d.name) > 27 {
-		return fmt.Errorf("Device name too long, max. 27 characters")
+	if instConf.Type() == instancetype.VM && len(d.name) > 27 {
+		return fmt.Errorf("Disk device name too long, max. 27 characters")
```

Comment:
Should we change this to something like `Disk device name too long for VM instance, max. 27 characters` so that when its applied in a profile to a mixture of instances, the error is clear which instance type its applying to.

Reply:
I'm not sure that it really matters as whether applied directly to a VM or to a profile which applies to one or more VMs, you'll need to shorten it anyway.

Reply:
Yes but when applying via a profile to a mixture of instances (i.e after they've been created) it may seem odd to the user that only their VMs are generating this error and not their containers.

Reply:
I think ultimately we may want to do two things to avoid this check entirely:
 - Put a cap at 64 chars for all device names (and document it), that should avoid some similar cases of this problem
 - For the VM disk case, fallback to the device ID if longer than what QMP can handle

I don't like this kind of special restriction that just applies to a particular device type on a particular instance type.

Reply:
Works for me. I don't like instance type specific restrictions either, which is why it was applied to all types originally.

Reply:
> I don't like this kind of special restriction that just applies to a particular device type on a particular instance type.

Okay, but it *only* apply to a *particular device on a particular instance type*. So it's better to warn the user.

Reply:
> * For the VM disk case, fallback to the device ID if longer than what QMP can handle

Only thing we need to consider is the `lxd recover` whilst the VM instance is running.

Reply:
> Put a cap at 64 chars for all device names.
It's also a big restriction, you can't interpolate the size people needs.

Reply:
> It's also a big restriction, you can't interpolate the size people needs.

I'm sure we'll break some people and I'm fine with it.
Not having any limit is an issue as it makes it impossible to validate that things will actually work on our side.

@tomponline The 64 chars length comes from something being a boundary size that people may expect along with in general allowing for dev name size + separators + project name being less than 108 chars which is our normal lower boundary for path length (socket max path).

Reply:
@stgraber shall I take this on?

Reply:
Testing an implementation of that